### PR TITLE
Add a command to highlight users in bot html pages

### PR DESCRIPTION
**Pull request by mia-pi-git**
https://www.smogon.com/forums/threads/add-highlights-to-bot-pages.3669880/
Command name is tentative. Didn't like it, but couldn't think of a better one.

### CODE REVIEW

**Comment by Zarel** [comment]
"requested the page recently, otherwise the first connection in the list" is rather different from "only if they have the page open", isn't it? Am I missing something?

**Comment by mia-pi-git** [comment]
Leaving this here too.

> "requested the page recently, otherwise the first connection in the list" is rather different from "only if they have the page open", isn't it? Am I missing something?

> [10:31:32] %mia pi: the last request page thing is only a way to figure out what connection to send it to
> [10:31:55] %mia pi: the highlight will only trigger if they have the page open since it's being sent to "view-bot-pageid-userid"
> [10:32:55] %mia pi: it's the same as how /ht list works

**Comment by sparkychildcharlie** [comment]
Is it just me, or is there no permission check for this command.  As in users can abuse it to highlight spam other people once they figure out who has what page open? (or just randomly send ``|tempnotify|`` messages to pretty much everyone on the server - idk if that would still hl if room is nonexistent, but yeah...)

**Comment by Zarel** [comment]
Hm. We might want to worry about notification spam.

If the client has separate "notification" and "bot notification" settings, that might help.

**Comment by mia-pi-git** [comment]
I mean, if you don't want to be spammed, just close the page? I would also assume most bot owners who own a bot with the power to do this, would know not to.

**Comment by Zarel** [comment]
I think it's pretty common to want to use a page but also be annoyed by notification spam.

...I might even make bot notifications opt-in. That's probably a good idea, actually. Put it in user settings?

File: server/chat-commands/admin.ts [hunk 1]
```diff
@@ -235,6 +235,44 @@ export const commands: ChatCommands = {
...
+		if (!userid || !pageid || !target) return this.parse(`/help highlighthtmlpage`);
+		const targetUser = Users.get(userid);
+		if (!targetUser || !targetUser.connected) {
+			throw new Chat.ErrorMessage(`User ${this.targetUsername} is not currently online.`);
+		}
+		if (targetUser.locked && !this.user.can('lock')) {
+			throw new Chat.ErrorMessage("This user is currently locked, so you cannot send them highlights.");
+		}
+
+		const buf = `|tempnotify|bot-${pageid}|A bot has notified you!|${highlight ? highlight : ''}`;
```
Comment (Zarel) [inline server/chat-commands/admin.ts:254]:
```suggestion
		const buf = `|tempnotify|bot-${pageid}|A bot has notified you!|${highlight ? highlight : ''}`;
```

I think "A bot has notified you" is a pretty ugly notification message. Why not let the bot set it? Maybe add `[from ${botName}]` or something before or after it to be clearer.

**Comment by Zarel** [comment]
This commit log is hella messy; did you mess up a merge?

**Comment by mia-pi-git** [comment]
> This commit log is hella messy; did you mess up a merge?

Yes, apparently. I'll handle it when I'm back on a computer.

**Comment by mia-pi-git** [comment]
Fixed, @Zarel.

**Comment by Zarel** [comment]
I would still prefer if this was opt-in, to be clear. But I won't block on that.

**Pull request merged**